File: KLine-Chart-OC/KLine-Chart-OC/KLineChart/Renderer/MainChartRenderer.m
//
//  MainChartRenderer.m
//  KLine-Chart-OC
//
//  Created by 何俊松 on 2020/3/10.
//  Copyright © 2020 hjs. All rights reserved.
//

#import "MainChartRenderer.h"
#import "ChartStyle.h"
#import "NSString+Rect.h"
@interface MainChartRenderer()
@property(nonatomic,assign) BOOL isLine;
@property(nonatomic,assign) CGFloat contentPadding;
@property(nonatomic,assign) MainState state;

@end


@implementation MainChartRenderer

- (instancetype)initWithMaxValue:(CGFloat)maxValue minValue:(CGFloat)minValue chartRect:(CGRect)chartRect candleWidth:(CGFloat)candleWidth topPadding:(CGFloat)topPadding isLine:(BOOL)isLine state:(MainState)state {
    if (self = [super initWithMaxValue:maxValue minValue:minValue chartRect:chartRect candleWidth:candleWidth topPadding:topPadding]) {
        self.contentPadding = 20;
        self.isLine = isLine;
        self.state = state;
        CGFloat diff = maxValue - minValue;
        CGFloat newscaly = 1;
        CGFloat newDiff = 0;
        CGFloat value = 0;
        if(diff != 0) {
            newscaly = (chartRect.size.height - _contentPadding)/ diff;
            newDiff = chartRect.size.height / newscaly;
            value = (newDiff - diff) / 2;
        }
        if(newDiff > diff) {
            self.scaleY = newscaly;
            self.maxValue += value;
            self.minValue -= value;
        }
    }
    return self;
}

- (void)drawGrid:(CGContextRef)context gridRows:(NSUInteger)gridRows gridColums:(NSUInteger)gridColums {
    CGContextSetStrokeColorWithColor(context, ChartColors_gridColor.CGColor);
    CGContextSetLineWidth(context, 0.5);
    CGFloat columsSpace = self.chartRect.size.width / (CGFloat)(gridColums);
    for (int index = 0;  index < gridColums; index++) {
        CGContextMoveToPoint(context, index * columsSpace, 0);
        CGContextAddLineToPoint(context, index * columsSpace, CGRectGetMaxY(self.chartRect));
        CGContextDrawPath(context, kCGPathFillStroke);
    }
    CGFloat rowSpace = self.chartRect.size.height / (CGFloat)gridRows;
    for (int index = 0;  index < gridRows; index++) {
         CGContextMoveToPoint(context, 0, index * rowSpace + ChartStyle_topPadding);
         CGContextAddLineToPoint(context, CGRectGetMaxX(self.chartRect), index * rowSpace + ChartStyle_topPadding);
         CGContextDrawPath(context, kCGPathFillStroke);
     }
    CGContextAddRect(context, self.chartRect);
    CGContextDrawPath(context, kCGPathStroke);
}

- (void)drawChart:(CGContextRef)context lastPoit:(KLineModel *)lastPoint curPoint:(KLineModel *)curPoint curX:(CGFloat)curX {
    if(!_isLine) {
        [self drawCandle:context curPoint:curPoint curX:curX];
    }
    if (lastPoint != nil) {
        if(_isLine) {
            [self drawKLine:context lastValue:lastPoint.close curValue:curPoint.close curX:curX];
        } else if (_state == MainStateMA) {
            [self drawMaLine:context lastPoit:lastPoint curPoint:curPoint curX:curX];
        } else if (_state == MainStateBOLL) {
            [self drawBollLine:context lastPoit:lastPoint curPoint:curPoint curX:curX];
        }
    }
}

- (void)drawKLine:(CGContextRef)context lastValue:(CGFloat)lastValue curValue:(CGFloat)curValue curX:(CGFloat)curX  {
    CGFloat x1 = curX;
    CGFloat y1 = [self getY:curValue];
    CGFloat x2 = curX + self.candleWidth + ChartStyle_canldeMargin;
    CGFloat y2 = [self getY:lastValue];
    CGContextSetLineWidth(context, 1);
    CGContextSetStrokeColorWithColor(context, ChartColors_kLineColor.CGColor);
    CGContextMoveToPoint(context, x1, y1);
    CGContextAddCurveToPoint(context, (x1 + x2) / 2.0, y1,  (x1 + x2) / 2.0, y2, x2, y2);
    CGContextDrawPath(context, kCGPathFillStroke);
    
    CGMutablePathRef path = CGPathCreateMutable();
    CGPathMoveToPoint(path, &CGAffineTransformIdentity, x1, CGRectGetMaxY(self.chartRect));
    CGPathAddLineToPoint(path, &CGAffineTransformIdentity, x1, y1);
    CGPathAddCurveToPoint(path, &CGAffineTransformIdentity, (x1 + x2) / 2.0, y1, (x1 + x2) / 2.0, y2, x2, y2);
    CGPathAddLineToPoint(path,  &CGAffineTransformIdentity, x2, CGRectGetMaxY(self.chartRect));
    CGPathCloseSubpath(path);
    CGContextAddPath(context, path);
    
    CGContextClip(context);
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    CGFloat locations[] = {0,1};
    NSArray *colors = @[(__bridge id)[UIColor rgb_r:0x4c g:0x86 b:0xCD alpha:1].CGColor, (__bridge id)[UIColor rgb_r:0x00 g:0x00 b:0x00 alpha:0].CGColor];
    CGGradientRef gradient = CGGradientCreateWithColors(colorSpace, (CFArrayRef) colors, locations);
    CGColorSpaceRelease(colorSpace);
    CGPoint start = CGPointMake((x1 + x2) / 2, CGRectGetMinY(self.chartRect));
    CGPoint end = CGPointMake((x1 + x2) / 2, CGRectGetMaxY(self.chartRect));
    CGContextDrawLinearGradient(context, gradient, start, end, 0);
    CGContextResetClip(context);
    
    CGColorSpaceRelease(colorSpace);
    CGPathRelease(path);
    CGGradientRelease(gradient);
    
}

- (void)drawMaLine:(CGContextRef)context lastPoit:(KLineModel *)lastPoint curPoint:(KLineModel *)curPoint curX:(CGFloat)curX {
    if(curPoint.MA5Price != 0) {
        [self drawLine:context lastValue:lastPoint.MA5Price curValue:curPoint.MA5Price curX:curX color:ChartColors_ma5Color];
    }
    if(curPoint.MA10Price != 0) {
           [self drawLine:context lastValue:lastPoint.MA10Price curValue:curPoint.MA10Price curX:curX color:ChartColors_ma10Color];
    }
    if(curPoint.MA30Price != 0) {
           [self drawLine:context lastValue:lastPoint.MA30Price curValue:curPoint.MA30Price curX:curX color:ChartColors_ma30Color];
    }
}


- (void)drawBollLine:(CGContextRef)context lastPoit:(KLineModel *)lastPoint curPoint:(KLineModel *)curPoint curX:(CGFloat)curX {
    if(curPoint.up != 0) {
        [self drawLine:context lastValue:lastPoint.up curValue:curPoint.up curX:curX color:ChartColors_ma5Color];
    }
    if(curPoint.mb != 0) {
           [self drawLine:context lastValue:lastPoint.mb curValue:curPoint.mb curX:curX color:ChartColors_ma10Color];
    }
    if(curPoint.dn != 0) {
           [self drawLine:context lastValue:lastPoint.dn curValue:curPoint.dn curX:curX color:ChartColors_ma30Color];
    }
}


- (void)drawCandle:(CGContextRef)context curPoint:(KLineModel *)curPoint curX:(CGFloat)curX {
    CGFloat high = [self getY:curPoint.high];
    CGFloat low = [self getY:curPoint.low];
    CGFloat open = [self getY:curPoint.open];
    CGFloat close = [self getY:curPoint.close];
    UIColor *color = ChartColors_dnColor;
    
    if(open > close) {
        color = ChartColors_upColor;
    }
    
    CGContextSetStrokeColorWithColor(context, color.CGColor);
    CGContextSetLineWidth(context, ChartStyle_candleLineWidth);
    CGContextMoveToPoint(context, curX, high);
    CGContextAddLineToPoint(context, curX, low);
    CGContextDrawPath(context, kCGPathFillStroke);
    
    CGContextSetStrokeColorWithColor(context, color.CGColor);
    CGContextSetLineWidth(context, self.candleWidth);
    CGContextMoveToPoint(context, curX, open);
    CGContextAddLineToPoint(context, curX, close);
    CGContextDrawPath(context, kCGPathFillStroke);
    
}

- (void)drawTopText:(CGContextRef)context curPoint:(KLineModel *)curPoint {
    NSMutableAttributedString *topAttributeText = [[NSMutableAttributedString alloc] init];
    if(curPoint.MA5Price != 0) {
        NSString *str = [NSString stringWithFormat:@"MA5:%.2f   ",curPoint.MA5Price];
        NSAttributedString *attr = [[NSAttributedString alloc] initWithString:str attributes:@{NSFontAttributeName: [UIFont systemFontOfSize:ChartStyle_defaultTextSize],NSForegroundColorAttributeName: ChartColors_ma5Color}];
        [topAttributeText appendAttributedString:attr];
    }
    if(curPoint.MA10Price != 0) {
        NSString *str = [NSString stringWithFormat:@"MA10:%.2f    ",curPoint.MA10Price];
        NSAttributedString *attr = [[NSAttributedString alloc] initWithString:str attributes:@{NSFontAttributeName: [UIFont systemFontOfSize:ChartStyle_defaultTextSize],NSForegroundColorAttributeName: ChartColors_ma10Color}];
        [topAttributeText appendAttributedString:attr];
    }
    if(curPoint.MA30Price != 0) {
        NSString *str = [NSString stringWithFormat:@"MA30:%.2f   ",curPoint.MA30Price];
        NSAttributedString *attr = [[NSAttributedString alloc] initWithString:str attributes:@{NSFontAttributeName: [UIFont systemFontOfSize:ChartStyle_defaultTextSize],NSForegroundColorAttributeName: ChartColors_ma30Color}];
        [topAttributeText appendAttributedString:attr];
    }
    [topAttributeText drawAtPoint:CGPointMake(5, 6)];
}

- (void)drawRightText:(CGContextRef)context gridRows:(NSUInteger)gridRows gridColums:(NSUInteger)gridColums {
    CGFloat rowSpace = self.chartRect.size.height / (CGFloat)gridRows;
    for (int i = 0; i <= gridRows; i++) {
        CGFloat position = 0;
        position = (CGFloat)(gridRows - i) * rowSpace;
        CGFloat value = position / self.scaleY + self.minValue;
        NSString *valueStr = [NSString stringWithFormat:@"%.2f",value];
        CGRect rect = [valueStr getRectWithFontSize:ChartStyle_reightTextSize];
        CGFloat y = 0;
        if(i == 0) {
            y = [self getY:value];
        } else {
            y = [self getY:value] - rect.size.height;
        }
        [self drawText:valueStr atPoint:CGPointMake(self.chartRect.size.width - rect.size.width, y) fontSize:ChartStyle_reightTextSize textColor:ChartColors_reightTextColor];
    }
}


- (CGFloat)getY:(CGFloat)value {
    return self.scaleY * (self.maxValue - value) + CGRectGetMinY(self.chartRect);
}

@end

File: KLine-Chart/KLine-Chart/KLineChart/Renderer/MainChartRenderer.swift
//
//  MainChartRenderer.swift
//  KLine-Chart
//
//  Created by 何俊松 on 2020/3/1.
//  Copyright © 2020 hjs. All rights reserved.
//

import UIKit

class MainChartRenderer: BaseChartRenderer {

    var contentPadding: CGFloat = 20
    
    var isLine: Bool = false
    var state: MainState = .none
    
    init(maxValue: CGFloat,minValue: CGFloat,chartRect: CGRect,candleWidth: CGFloat,topPadding: CGFloat,isLine: Bool,state: MainState) {
        super.init(maxValue: maxValue, minValue: minValue, chartRect: chartRect, candleWidth: candleWidth,topPadding: topPadding)
        self.isLine = isLine
        self.state = state
        let diff = maxValue - minValue
        print(chartRect.height)
        let newScalY = (chartRect.height - contentPadding) / diff
        let newDiff = chartRect.height / newScalY
        let value = (newDiff - diff) / 2
        if newDiff > diff {
            scaleY = newScalY
            self.maxValue += value
            self.minValue -= value
        }
    }
    
    override func drawGrid(context: CGContext, gridRows: Int, gridColums: Int) {
        context.setStrokeColor(ChartColors.gridColor.cgColor)
        context.setLineWidth(0.5)
        let columsSpace = chartRect.width / CGFloat(gridColums)
        for index in 0..<gridColums {
            context.move(to: CGPoint(x: CGFloat(index) * columsSpace, y: 0))
            context.addLine(to: CGPoint(x: CGFloat(index) * columsSpace, y: chartRect.maxY))
            context.drawPath(using: CGPathDrawingMode.fillStroke)
        }
         let rowSpace = chartRect.height / CGFloat(gridRows)
        for index in 0...gridRows {
            context.move(to: CGPoint(x: 0, y: CGFloat(index) * rowSpace + ChartStyle.topPadding))
            context.addLine(to: CGPoint(x: chartRect.maxX, y: CGFloat(index) * rowSpace + ChartStyle.topPadding))
            context.drawPath(using: CGPathDrawingMode.fillStroke)
        }
    }
    
    
    override func drawChart(context: CGContext, lastPoint: KLineModel?, curPoint: KLineModel, curX: CGFloat) {
        if(!isLine) {
            drawCandle(context: context, curPoint: curPoint, curX: curX)
        }
        if let _lastPoint = lastPoint {
            if(isLine) {
                drawKLine(context: context, lastValue: _lastPoint.close, curValue: curPoint.close, curX: curX)
            } else if state == MainState.ma {
                drawMaLine(context: context, lastPoint: _lastPoint, curPoint: curPoint, curX: curX)
            } else if state == MainState.boll {
                drawBOLL(context: context, lastPoint: _lastPoint, curPoint: curPoint, curX: curX)
            }
        }
    }
    
    func drawKLine(context: CGContext,lastValue: CGFloat, curValue: CGFloat, curX: CGFloat) {
       let x1 = curX
       let y1 = getY(curValue)
       let x2 = curX + candleWidth + ChartStyle.canldeMargin
       let y2 = getY(lastValue)
        context.setLineWidth(1)
        context.setStrokeColor(ChartColors.kLineColor.cgColor)
        context.move(to: CGPoint(x: x1, y: y1))
        context.addCurve(to: CGPoint(x: x2, y: y2), control1: CGPoint(x: (x1 + x2) / 2, y: y1), control2: CGPoint(x: (x1 + x2) / 2, y: y2))
        context.drawPath(using: CGPathDrawingMode.fillStroke)
        
        //创建并设置路径
        let path = CGMutablePath()
        path.move(to: CGPoint(x: x1, y: chartRect.maxY))
        path.addLine(to: CGPoint(x: x1, y: y1))
        path.addCurve(to: CGPoint(x: x2, y: y2), control1: CGPoint(x: (x1 + x2) / 2, y: y1), control2: CGPoint(x: (x1 + x2) / 2, y: y2))
        path.addLine(to: CGPoint(x: x2, y: chartRect.maxY))
        path.closeSubpath()
        //添加路径到图形上下文
        context.addPath(path)
        context.clip()
        
        //使用rgb颜色空间
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        //颜色数组（这里使用三组颜色作为渐变）fc6820
        let compoents:[CGFloat] = [0x4C/255, 0x86/255, 0xCD/255, 0x55/255,
                                   0x00/255, 0x00/255, 0x00/255, 0]
        //没组颜色所在位置（范围0~1)
        let locations:[CGFloat] = [0,1]
        //生成渐变色（count参数表示渐变个数）
        let gradient = CGGradient(colorSpace: colorSpace, colorComponents: compoents,
                                  locations: locations, count: locations.count)!
         
        //渐变开始位置
        let start = CGPoint(x: (x1 + x2) / 2, y: chartRect.minY)
        //渐变结束位置
        let end = CGPoint(x: (x1 + x2) / 2, y: chartRect.maxY)
        //绘制渐变
        context.drawLinearGradient(gradient, start: start, end: end,
                                   options: .drawsAfterEndLocation)
        context.resetClip()

        
    }
    
    func drawMaLine(context: CGContext,lastPoint: KLineModel, curPoint: KLineModel,curX: CGFloat) {
        if(curPoint.MA5Price != 0) {
            drawLine(context: context, lastValue: lastPoint.MA5Price, curValue: curPoint.MA5Price, curX: curX, color: ChartColors.ma5Color)
        }
        if(curPoint.MA10Price != 0) {
            drawLine(context: context, lastValue: lastPoint.MA10Price, curValue: curPoint.MA10Price, curX: curX, color: ChartColors.ma10Color)
        }
        if(curPoint.MA30Price != 0) {
            drawLine(context: context, lastValue: lastPoint.MA30Price, curValue: curPoint.MA30Price, curX: curX, color: ChartColors.ma30Color)
        }
    }
    
    func drawBOLL(context: CGContext,lastPoint: KLineModel, curPoint: KLineModel,curX: CGFloat) {
        if(curPoint.up != 0) {
            drawLine(context: context, lastValue: lastPoint.up, curValue: curPoint.up, curX: curX, color: ChartColors.ma5Color)
        }
        if(curPoint.mb != 0) {
            drawLine(context: context, lastValue: lastPoint.mb, curValue: curPoint.mb, curX: curX, color: ChartColors.ma10Color)
        }
        if(curPoint.dn != 0) {
            drawLine(context: context, lastValue: lastPoint.dn, curValue: curPoint.dn, curX: curX, color: ChartColors.ma30Color)
        }
    }
    
    func drawCandle(context: CGContext, curPoint: KLineModel,curX: CGFloat) {
        let high = getY(curPoint.high)
        let low = getY(curPoint.low)
        let open = getY(curPoint.open)
        let close = getY(curPoint.close)
         var  color = ChartColors.dnColor
        if open > close {
           color = ChartColors.upColor
        }
        context.setStrokeColor(color.cgColor)
        context.setLineWidth(ChartStyle.candleLineWidth)
        context.move(to: CGPoint(x: curX , y: high))
        context.addLine(to: CGPoint(x: curX, y: low))
        context.drawPath(using: CGPathDrawingMode.fillStroke)
        context.setStrokeColor(color.cgColor)
        context.setLineWidth(candleWidth)
        context.move(to: CGPoint(x: curX, y: open))
        context.addLine(to: CGPoint(x: curX, y: close))
        context.drawPath(using: CGPathDrawingMode.fillStroke)
    }
    
    override func drawRightText(context: CGContext, gridRows: Int, gridColums: Int) {
        let rowSpace = chartRect.height / CGFloat(gridRows)
        for i in 0...gridRows {
            var  position: CGFloat = 0;
            position = CGFloat(gridRows - i) * rowSpace
            let value = position / scaleY + minValue
            let valueStr = String(format: "%.2f", value)
            let rect = calculateTextRect(text: valueStr, fontSize: ChartStyle.reightTextSize)
            var y: CGFloat = 0
            if i == 0 {
                y = getY(value)
            } else {
                y = getY(value) - rect.height
            }
            valueStr.draw(at: CGPoint(x: chartRect.width - rect.width, y: y), withAttributes: [NSAttributedString.Key.font : UIFont.systemFont(ofSize: ChartStyle.reightTextSize), NSAttributedString.Key.foregroundColor: ChartColors.reightTextColor])
        }
    }
    
    override func drawTopText(context: CGContext, curPoint: KLineModel) {
        let topAttributeText = NSMutableAttributedString()
        if curPoint.MA5Price != 0 {
            let ma5Price  = String(format: "%.2f", curPoint.MA5Price)
            let ma5Attr = NSAttributedString(string: "MA5:\(ma5Price)    ", attributes: [NSAttributedString.Key.font : UIFont.systemFont(ofSize: ChartStyle.defaultTextSize),NSAttributedString.Key.foregroundColor: ChartColors.ma5Color])
            topAttributeText.append(ma5Attr)
        }
        if curPoint.MA10Price != 0 {
            let ma10Price  = String(format: "%.2f", curPoint.MA5Price)
            let ma10Attr = NSAttributedString(string: "MA10:\(ma10Price)    ", attributes: [NSAttributedString.Key.font : UIFont.systemFont(ofSize: ChartStyle.defaultTextSize),NSAttributedString.Key.foregroundColor: ChartColors.ma10Color])
            topAttributeText.append(ma10Attr)
        }
        
        if curPoint.MA30Price != 0 {
            let ma30Price  = String(format: "%.2f", curPoint.MA5Price)
            let ma30Attr = NSAttributedString(string: "MA30:\(ma30Price)    ", attributes: [NSAttributedString.Key.font : UIFont.systemFont(ofSize: ChartStyle.defaultTextSize),NSAttributedString.Key.foregroundColor: ChartColors.ma30Color])
            topAttributeText.append(ma30Attr)
        }
        topAttributeText.draw(at: CGPoint(x: 5, y: 6))
    }
    
    override func getY(_ value: CGFloat) -> CGFloat {
        return scaleY * (maxValue - value) + chartRect.minY
    }
  
    
}
